pkg/nsx/services/compare.go

```diff
@@ -0,0 +1,118 @@
+package services
+
+import (
+	"encoding/json"
+	"sort"
+
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/data/serializers/cleanjson"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+)
+
+func SecurityPolicyEqual(existingSecurityPolicy *model.SecurityPolicy, securityPolicy *model.SecurityPolicy) bool {
+	s1, _ := json.Marshal(simplifySecurityPolicy(existingSecurityPolicy))
+	s2, _ := json.Marshal(simplifySecurityPolicy(securityPolicy))
+	if string(s1) == string(s2) {
+		return true
+	}
+	log.Info("security policy diff", "nsx sp", s1, "k8s sp", s2)
+	return false
+}
+
+func RulesEqual(existingRules []model.Rule, rules []model.Rule) (bool, []model.Rule) {
+	// sort the rules by id, otherwise expandRule may return different results, only the sequence of the
+	// rule is different, so sort by port number, and it avoids the needless updates.
+	var sortRules = func(rules []model.Rule) {
+		sort.Slice(rules, func(i, j int) bool {
+			return *(rules[i].Id) > *(rules[j].Id)
+		})
+	}
+	sortRules(existingRules)
+	sortRules(rules)
+
+	isEqual := true
+	// legacyRules means the rules that are not in the new rules, we should destroy them.
+	var legacyRules []model.Rule
+
+	diffIndex := len(existingRules) - len(rules)
+	if diffIndex > 0 {
+		isEqual = false
+		legacyRules = append(legacyRules, existingRules[diffIndex:]...)
+		return isEqual, legacyRules
+	} else if diffIndex < 0 {
+		isEqual = false
+		return isEqual, legacyRules
+	}
+
+	for i := 0; i < len(rules); i++ {
+		r1, _ := simplifyRule(&existingRules[i]).GetDataValue__()
```

Comment:
put 47-56 line to function like RuleEqual?

Reply:
Good